onnxruntime/core/providers/cpu/ml/tree_ensemble_aggregator.h

```diff
@@ -23,9 +23,20 @@ struct TreeNodeElementId {
   }
   struct hash_fn {
     std::size_t operator()(const TreeNodeElementId& key) const {
-      std::size_t h1 = std::hash<int64_t>()(key.tree_id);
-      std::size_t h2 = std::hash<int64_t>()(key.node_id);
-      return h1 ^ h2;
+      // Employs Elegant Pairing function to produce a unique non-negative integer from two non-negative integers.
+      // Reference: http://szudzik.com/ElegantPairing.pdf
+
+      u_int64_t combined_id;
+      // Use absolute value, before casting, to reduce overflow chance if any of the values is negative.
+      // In the worst (and also unlike) case, we will map 4 pairs of tree_id and node_id to the same value
+      u_int64_t x = static_cast<u_int64_t>(std::abs(key.tree_id));
```

Comment:
Isn't the overflow exactly what you want here, though?

Reply:
Yes, that is a good point. I asked myself the same question. However, if we ignore the overflow caused by casting a signed integer to an unsigned one we would have a decrease in the spread of values into different buckets. By using `abs` we create a more "predictable" scenario.

To test it, I created a small program to compare all three hash functions (the current one, Elegant Pairing with `abs` and Elegant Pairing without `abs`). It generates hash values for each pair of integers inside the interval. Those were the results:

## [-10000, 10000] Interval (mixing negatives and non-negatives numbers)

### Elegant Pairing with "abs"
```
[-10000, 10000] [-10000, 10000] 1 1

ID's pair count: 400040001
distinct hash values produced: 100020001
colision count: 400040000

there are 100000000 buckets with 4 elements
there are 20000 buckets with 2 elements
there are 1 buckets with 1 elements
```

### Elegant Pairing without "abs"
```
[-10000, 10000] [-10000, 10000] 0 1

ID's pair count: 400040001
distinct hash values produced: 100030001
collision count: 400030000

there are 33336668 buckets with 2 elements
there are 33336668 buckets with 3 elements
there are 13334668 buckets with 4 elements
there are 6667336 buckets with 5 elements
there are 3809907 buckets with 6 elements
...
there are 142 buckets with 279 elements
there are 141 buckets with 278 elements
there are 141 buckets with 280 elements
there are 141 buckets with 276 elements
there are 120 buckets with 282 elements
printed first and last 5 buckets with more than one element (still 272 remaining of 282)
```

### Current implementation
```
[-10000, 10000] [-10000, 10000] 0 0

ID's pair count: 400040001
distinct hash values produced: 32768
collision count: 400040001

there are 16384 buckets with 7234 elements
there are 12288 buckets with 16384 elements
there are 3584 buckets with 19522 elements
there are 448 buckets with 19968 elements
there are 32 buckets with 19970 elements
there are 31 buckets with 20000 elements
there are 1 buckets with 20001 elements
```

## [2^60, 2^60 + 5000] Interval (inevitable overflow)
Note that even in a scenario where overflows are inevitable (if ID's are big), Elegant Pairing is still able to better spread the values into different buckets.

### Elegant Pairing with "abs"
```
[1152921504606846976, 1152921504606851976] [1152921504606846976, 1152921504606851976] 1 1

ID's pair count: 25010001
distinct hash values produced: 25010001
collision count: 0

there are 25010001 buckets with 1 elements
```

### Current implementation
```
[1152921504606846976, 1152921504606851976] [1152921504606846976, 1152921504606851976] 0 0

ID's pair count: 25010001
distinct hash values produced: 8192
collision count: 25010001

there are 4096 buckets with 1810 elements
there are 3072 buckets with 4096 elements
there are 896 buckets with 4882 elements
there are 112 buckets with 4992 elements
there are 8 buckets with 4994 elements
there are 7 buckets with 5000 elements
there are 1 buckets with 5001 elements
```

Reply:
Source code used to produce the results reported above:
```c++
#include <cstdio>
#include <cstdint>
#include <cstring>
#include <unordered_map>
#include <queue>
#include <vector>
#include <map>

using namespace std;

// g++ -O2 -std=c++20 elegant_paring.cc

size_t elegant_pairing_hash(int64_t x0, int64_t y0, bool use_abs) {
  u_int64_t combined_id;

  u_int64_t x;
  u_int64_t y; 

  if (use_abs) {
    x = static_cast<u_int64_t>(abs(x0));
    y = static_cast<u_int64_t>(abs(y0));
  } else {
    x = static_cast<u_int64_t>(x0);
    y = static_cast<u_int64_t>(y0);
  }

  if (x >= y) {
    combined_id = x * x + (x + y);
  } else {
    combined_id = y * y + x;
  }
  return hash<u_int64_t>()(combined_id);
}

size_t current_hash(int64_t x0, int64_t y0) {
  size_t h1 = hash<int64_t>()(x0);
  size_t h2 = hash<int64_t>()(y0);
  return h1 ^ h2;
}

int main(int argc, char** argv) {
  if (argc != 7) {
    printf("[x_begin, x_end] [y_begin, y_end] hash_name use_abs\n");
  }

  int64_t x_begin;
  int64_t x_end;
  int64_t y_begin;
  int64_t y_end;

  sscanf(argv[1], "%lld", &x_begin);
  sscanf(argv[2], "%lld", &x_end);
  sscanf(argv[3], "%lld", &y_begin);
  sscanf(argv[4], "%lld", &y_end);

  bool use_elegant_pairing_hash = strcmp(argv[5], "true") == 0;
  bool use_abs = strcmp(argv[6], "true") == 0;

  printf("[%lld, %lld] [%lld, %lld] %d %d\n\n", x_begin, x_end, y_begin, y_end, use_abs, use_elegant_pairing_hash);

  if (x_begin > x_end ) {
    swap(x_begin, x_end);
  }

  if (y_begin > y_end ) {
    swap(y_begin, y_end);
  }

  unordered_map<size_t, u_int64_t> count_by_hash_value;
  u_int64_t pair_count = 0;
  u_int64_t colision_count = 0;

  for (int64_t x = x_begin; x <= x_end; x++) {
    for (int64_t y = y_begin; y <= y_end; y++) {
      std:size_t hash;
      if (use_elegant_pairing_hash) {
        hash = elegant_pairing_hash(x, y, use_abs);  
      } else {
        hash = current_hash(x, y);
      }

      if (count_by_hash_value.contains(hash)) {
        count_by_hash_value[hash] = count_by_hash_value[hash] + 1;
      
      } else {
        count_by_hash_value[hash] = 1;
      }
      
      //printf("%lu (%lX) <- %lld %lld\n", hash, hash, x, y);

      pair_count++;
    }
  }

  u_int64_t max = 0;
  map<size_t, u_int64_t> bucket_count_by_size;

  for (const pair<u_int64_t, size_t>& pair : count_by_hash_value) {
    u_int64_t bucket_size = pair.second;
    if (bucket_size > 1) {
      colision_count += bucket_size;
    }
    if (bucket_size > max) {
      max = bucket_size;
    }

    if (bucket_count_by_size.contains(bucket_size)) {
      bucket_count_by_size[bucket_size] = bucket_count_by_size[bucket_size] + 1;
    } else {
      bucket_count_by_size[bucket_size] = 1;
    }
  }
  printf("ID's pair count: %llu\n", pair_count);
  printf("distinct hash values produced: %lu\n", count_by_hash_value.size());
  printf("collision count: %llu\n", colision_count);
  printf("\n");

  vector<pair<size_t, u_int64_t>> sorted_bucket_count_by_size;
  for (auto& it : bucket_count_by_size) {
    sorted_bucket_count_by_size.push_back(it);
  }
 
   sort(sorted_bucket_count_by_size.begin(), sorted_bucket_count_by_size.end(), [&](auto a, auto b) {return a.second > b.second;});

  constexpr size_t bucket_limit = 5;
  size_t bucket_count = 0;
  for (const auto& [bucket_size, count] : sorted_bucket_count_by_size) {
    if (bucket_count < bucket_limit || sorted_bucket_count_by_size.size() - bucket_limit <= bucket_count) {
      if (bucket_limit * 2 < bucket_count_by_size.size() && sorted_bucket_count_by_size.size() - bucket_limit == bucket_count) {
        printf("...\n");
      }
      printf("there are %llu buckets with %lu elements\n", count, bucket_size);
    }
    ++bucket_count;
  }
  if (bucket_limit * 2 < bucket_count_by_size.size()) {
    printf("printed first and last %lu buckets with more than one element (still %lu remaining of %lu)\n", bucket_limit, bucket_count_by_size.size() - bucket_limit * 2, bucket_count_by_size.size());      
  }

  return 0;
}

```

Reply:
Thanks for the very detailed follow up!

Reply:
I am leaving this comment open since it seems a good discussion for future readers.